### PR TITLE
sql: correct lateral scoping in the face of outer joins

### LINKED ISSUE #6875: sql: incorrect outer join parsing
### What version of Materialize are you using?

```
$ materialized -v
materialized v0.7.4-dev (19c296dcd)
```

### How did you install Materialize?

* [ ] Docker image
* [ ] Linux release tarball
* [ ] APT package
* [ ] macOS release tarball
* [ ] Homebrew tap
* [x] Built from source

### What was the issue?

The following query

```
materialize=> explain raw plan for select * from
    foo,                       
    bar left join baz on y = z;
;
              Raw Plan              
------------------------------------
 %0 =                              +
 | Get materialize.public.foo (u9) +
                                   +
 %1 =                              +
 | Get materialize.public.bar (u11)+
                                   +
 %2 =                              +
 | InnerJoin %0 %1 on true         +
                                   +
 %3 =                              +
 | Get materialize.public.baz (u13)+
                                   +
 %4 =                              +
 | LeftOuterJoin %2 %3 on (#1 = #2)+
(1 row)
materialize=> 
```

should be syntactically equivalent to this other one:

```
materialize=> explain raw plan for select * from
    foo,
    (select * from bar left join baz on y = z);
;
              Raw Plan              
------------------------------------
 %0 =                              +
 | Get materialize.public.foo (u9) +
                                   +
 %1 =                              +
 | Get materialize.public.bar (u11)+
                                   +
 %2 =                              +
 | Get materialize.public.baz (u13)+
                                   +
 %3 =                              +
 | LeftOuterJoin %1 %2 on (#0 = #1)+
                                   +
 %4 =                              +
 | InnerJoin %0 %3 on true         +
(1 row)
materialize=> 
```

However, they result in different raw plan, where the joins happen in different order. The right plan should be the one in the second query, since the comma operator has lower precedence.

This also leads to a name resolution issue whereby the sibling tables in the comma join are seen from the ON clause of the outer join, where they shouldn't be seen. Hence, the following query is accepted while it should fail:

```
materialize=> explain select * from
    foo,
    bar left join baz on foo.y = baz.x;
                Optimized Plan                
----------------------------------------------
 %0 =                                        +
 | Get materialize.public.foo (u412)         +
 | ArrangeBy ()                              +
                                             +
 %1 =                                        +
 | Get materialize.public.bar (u414)         +
                                             +
 %2 = Let l0 =                               +
 | Join %0 %1                                +
 | | implementation = Differential %1 %0.()  +
 | | demand = (#0..#3)                       +
                                             +
 %3 =                                        +
 | Get %2 (l0)                               +
 | Filter !(isnull(#1))                      +
 | ArrangeBy (#1)                            +
                                             +
 %4 =                                        +
 | Get materialize.public.baz (u416)         +
 | Filter !(isnull(#0))                      +
                                             +
 %5 = Let l1 =                               +
 | Join %3 %4 (= #1 #4)                      +
 | | implementation = Differential %4 %3.(#1)+
 | | demand = (#0..#3, #5)                   +
                                             +
 %6 =                                        +
 | Get %5 (l1)                               +
 | Distinct group=(#1)                       +
 | ArrangeBy (#0)                            +
                                             +
 %7 =                                        +
 | Join %2 %6 (= #1 #4)                      +
 | | implementation = Differential %2 %6.(#0)+
 | | demand = (#0..#3)                       +
 | Negate                                    +
 | Project (#0..#3)                          +
                                             +
 %8 =                                        +
 | Union %7 %2                               +
 | Map null, null                            +
                                             +
 %9 =                                        +
 | Get %5 (l1)                               +
 | Project (#0..#3, #1, #5)                  +
                                             +
 %10 =                                       +
 | Union %8 %9                               +
 
(1 row)

materialize=> 
```

... compared to postgres:

```
asenac=# explain select * from
    foo,
    bar left join baz on foo.y = baz.x;
ERROR:  invalid reference to FROM-clause entry for table "foo"
LINE 3:     bar left join baz on foo.y = baz.x;
                                 ^
HINT:  There is an entry for table "foo", but it cannot be referenced from this part of the query.
asenac=# 
```

### PR DESCRIPTION
**Pull request by benesch**
Rework join planning so that we can handle lateral references that are
mixed with right and full outer joins. The implementation here largely
follows what @asenac laid out in @6875, except that the existing outer
scope design was found to be sufficient; that is, there is no concept of
a "sibling" scope introduced here.

The gist is that we now always plan joins with the RHS as a child scope
of the LHS. (Previously, we only created the new scope for joins if we
could eagerly notice that the RHS had a lateral element.) If the join is
a right or full join, we ban referencing any of the elements in the left
scope, as required by SQL:2008. When we encounter a non-lateral
subquery, we hide any lateral scope items entirely. Yes, it's weird that
non-lateral subqueries pretend lateral scope items don't exit, while
subqueries on the RHS of a right or full join *error* if they access a
lateral scope item, but that's what SQL demands.

~**WARNING:** The lowering code is amended to always treat inner and left joins as
correlated, even if they are not. This is correct, but it may produce
worse plans. I haven't investigated the movement of plans in much detail
yet.~

Fix #6875.

### Motivation

  * This PR fixes a recognized bug.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

**Comment by asenac** [comment]
This is brilliant. It's more or less what I tried to do back in the days, but much better and better organized :)

For the changes in the plans, as mentioned in slack, we could bring the old behavior back if re-introduced the old `lateral` flag as `correlated`, indicating whether the RHS of the join references any column from the LHS use that flag during lowering. This flag could be computed within `plan_join`. Alternatively, we could compute that on demand during lowering. The following patch is a quick and dirty implementation of that which leaves the plans as they were:

```
diff --git a/src/sql/src/plan/expr.rs b/src/sql/src/plan/expr.rs
index fecb56da7..aa87062db 100644
--- a/src/sql/src/plan/expr.rs
+++ b/src/sql/src/plan/expr.rs
@@ -805,6 +805,28 @@ impl HirRelationExpr {
         }
     }
 
+    pub fn is_correlated_join(&self) -> bool {
+        match self {
+            HirRelationExpr::Join {
+                left: _,
+                right,
+                on: _,
+                kind,
+            } if kind.is_lateral() => {
+                let mut correlated = false;
+                right.clone().visit_columns(0, &mut |depth, col| {
+                    if col.level > depth {
+                        if col.level - depth == 1 {
+                            correlated = true;
+                        }
+                    }
+                });
+                correlated
+            }
+            _ => false,
+        }
+    }
+
     pub fn project(self, outputs: Vec<usize>) -> Self {
         if outputs.iter().copied().eq(0..self.arity()) {
             // The projection is trivial. Suppress it.
diff --git a/src/sql/src/plan/lowering.rs b/src/sql/src/plan/lowering.rs
index 6df8c2000..875275387 100644
--- a/src/sql/src/plan/lowering.rs
+++ b/src/sql/src/plan/lowering.rs
@@ -296,7 +296,7 @@ impl HirRelationExpr {
                     right,
                     on,
                     kind,
-                } if kind.is_lateral() => {
+                } if self.is_correlated_join() => {
                     // A LATERAL join is a join in which the right expression has
                     // access to the columns in the left expression. It turns out
                     // this is *exactly* our branch operator, plus some additional
```

The code above would be acceptable if just added a non-mutable version of `visit_columns` method.

**Comment by asenac** [comment]
Note that the QGM tests need to be re-generated to catch up with the changes in the input HIR. That be done by running the following command within `src/sql` directory:

```
REWRITE=1 cargo test  -- test_hir_generator
```

Although I think we should tweak the HIR->QGM generation code so that we remove those noisy joins with the join identity at generation time.

**Comment by sploiselle** [comment]
Glad to cede review to @asenac; I'll need a day or two to parse through this.

**Review by asenac — COMMENTED** [review]
This is so good that it makes a task I spent quite a few days and failed at look very simple, even though it is not.

I'd fix the lowering regression before merging this though. At this point my preference would be to add an extra commit with a version of the patch mentioned in one of my previous comments.

File: src/sql/src/plan/scope.rs [hunk 1]
```diff
@@ -119,6 +127,26 @@ pub struct Scope {
...
+    // It's easiest to understand with an example. Consider this query:
+    //
+    //     SELECT (SELECT * FROM tab1, tab2, (SELECT tab1.a, tab3.a) FROM tab3, tab4
+    //     Scope 1:                          -----------------------
+    //     Scope 2:                    ----
+    //     Scope 3:              ----
+    //     Scope 4:                                                             ----
+    //     Scope 5:                                                       ----
+    //
+    // Note that the because the subquery is not marked `LATERAL`, its reference
```
[the remark this answers is not in the document]
Comment (benesch) [reply src/sql/src/plan/scope.rs]: Sure thing. Fixed.

File: src/sql/src/plan/scope.rs [hunk 2]
```diff
@@ -119,6 +127,26 @@ pub struct Scope {
     pub items: Vec<ScopeItem>,
     // items inherited from an enclosing query
     pub outer_scope: Option<Box<Scope>>,
+    // Whether this scope starts a new chain of lateral outer scopes.
+    //
+    // It's easiest to understand with an example. Consider this query:
+    //
+    //     SELECT (SELECT * FROM tab1, tab2, (SELECT tab1.a, tab3.a) FROM tab3, tab4
```
[the remark this answers is not in the document]
Comment (benesch) [reply src/sql/src/plan/scope.rs]: Good catch! Fixed.

**Review by benesch — COMMENTED** [review]
> I'd fix the lowering regression before merging this though. At this point my preference would be to add an extra commit with a version of the patch mentioned in one of my previous comments.

Done! I taught the planner to elide any no-op joins, and taught lowering to use the correlated join path only if the RHS is actually correlated. There are no longer any plan regressions, and in fact a few plan improvements because we're now a bit smarter about detecting no-op joins.

File: src/sql/src/plan/lowering.rs [hunk 3]
```diff
@@ -296,8 +296,8 @@ impl HirRelationExpr {
                     right,
                     on,
                     kind,
-                } if kind.is_lateral() => {
-                    // A LATERAL join is a join in which the right expression has
+                } if right.is_correlated() => {
```
Comment (asenac) [inline src/sql/src/plan/lowering.rs:299]: The commit message is no longer correct, since now lowering is smarter than it was.

**Review by asenac — APPROVED** [review]
Sweet!

File: src/sql/src/plan/lowering.rs (hunk 3 above)
Comment (benesch) [reply src/sql/src/plan/lowering.rs:299]: Thanks, fixed that an added a release note.

**Comment by benesch** [comment]
@asenac, thank you so much for all the help, and the super fast reviews! Really glad to be finally putting this issue to bed. 

> Glad to cede review to @asenac; I'll need a day or two to parse through this.

Ack, wanted to at least give ya a heads up that I had to tweak the `USING` priority stuff since I couldn't get the priorities to stick in the new (correct) join implementation. The bool that used to be called `visible_to_wildcard` does all the work now, under a new name `allow_unqualified_reference`.

**Pull request merged**